src/network/lan.cpp

```diff
@@ -0,0 +1,297 @@
+
+//copypaste from ../socket.cpp
+#ifdef _WIN32
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+// Without this some of the network functions are not found on mingw
+#ifndef _WIN32_WINNT
+#define _WIN32_WINNT 0x0501
+#endif
+#include <windows.h>
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#define LAST_SOCKET_ERR() WSAGetLastError()
+typedef SOCKET socket_t;
+typedef int socklen_t;
+#else
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <unistd.h>
+#include <arpa/inet.h>
+
+#ifndef __ANDROID__
+#include <ifaddrs.h>
+#define HAVE_IFADDRS 1
+#endif
+
+#define LAST_SOCKET_ERR() (errno)
+typedef int socket_t;
+#endif
+
+const static unsigned short int adv_port = 29998;
+static std::string ask_str;
+
+lan_adv::lan_adv() : Thread("lan_adv")
+{
+}
+
+void lan_adv::ask()
+{
+	if (!isRunning()) start();
+
+	if (ask_str.empty()) {
+		Json::Value j;
+		j["cmd"] = "ask";
+		ask_str = fastWriteJson(j);
+	}
+
+	send_string(ask_str);
+}
+
+void lan_adv::send_string(const std::string &str)
+{
+	try {
+		sockaddr_in addr = {};
+		addr.sin_family = AF_INET;
+		addr.sin_port = htons(adv_port);
+		addr.sin_addr.s_addr = htonl(INADDR_BROADCAST);
+		UDPSocket socket_send(false);
+		int set_option_on = 1;
+		setsockopt(socket_send.GetHandle(), SOL_SOCKET, SO_BROADCAST,
+				(const char *)&set_option_on, sizeof(set_option_on));
+		socket_send.Send(Address(addr), str.c_str(), str.size());
+	} catch (const std::exception &e) {
+		verbosestream << "udp broadcast send4 fail " << e.what() << "\n";
+	}
+
+	std::vector<uint32_t> scopes;
+// todo: windows and android
+
+#if HAVE_IFADDRS
+	struct ifaddrs *ifaddr = nullptr, *ifa = nullptr;
+	if (getifaddrs(&ifaddr) < 0) {
+	} else {
+		for (ifa = ifaddr; ifa; ifa = ifa->ifa_next) {
+			if (!ifa->ifa_addr)
+				continue;
+			if (ifa->ifa_addr->sa_family != AF_INET6)
+				continue;
+
+			auto sa = *((struct sockaddr_in6 *)ifa->ifa_addr);
+			if (sa.sin6_scope_id)
+				scopes.push_back(sa.sin6_scope_id);
+
+			/*errorstream<<"in=" << ifa->ifa_name << " a="<<Address(*((struct sockaddr_in6*)ifa->ifa_addr)).serializeString()<<" ba=" << ifa->ifa_broadaddr <<" sc=" << sa.sin6_scope_id <<" fl=" <<  ifa->ifa_flags
+			//<< " bn=" << Address(*((struct sockaddr_in6*)ifa->ifa_broadaddr)).serializeString()
+			<<"\n"; */
+		}
+	}
+	freeifaddrs(ifaddr);
+#endif
+
+	if (scopes.empty())
+		scopes.push_back(0);
+
+	struct addrinfo hints
+	{
+	};
+
+	hints.ai_family = AF_INET6;
+	hints.ai_socktype = SOCK_DGRAM;
+	hints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;
+	struct addrinfo *result = nullptr;
+	if (!getaddrinfo("ff02::1", nullptr, &hints, &result)) {
+		for (auto info = result; info; info = info->ai_next) {
+			try {
+				sockaddr_in6 addr = *((struct sockaddr_in6 *)info->ai_addr);
+				addr.sin6_port = htons(adv_port);
+				UDPSocket socket_send(true);
+				int set_option_on = 1;
+				setsockopt(socket_send.GetHandle(), SOL_SOCKET, SO_BROADCAST,
```

Comment:
better to use MULTICAST than broadcast for such discovery

Reply:
Yes, but I as of right now don't know how I would implement that, that is what @proller had in freeminer. If this needs to be changed then I will try to learn how I can implement it in myself if no one else is working on it.

Reply:
Do you or anyone else know how I could use multicast instead?